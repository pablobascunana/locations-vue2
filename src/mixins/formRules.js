import Vue from 'vue'

Vue.mixin({
  data() {
    return {
      formRules: {
        loginCredentials: true,
        required: value => !!value || this.$t('validations.required'),
        maxLength: value => value.length < 101 || this.$t('validations.maxLength'),
        validEmail: value => /(([^<>()[\]\\.,;:\s@"]+(\.[^<>()[\]\\.,;:\s@"]+)*)|(".+"))@(([[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\])|(([a-zA-Z\-0-9]+\.)+[a-zA-Z]{2,}))$/.test(value) || this.$t('validations.validEmail'),
        samePassword: value => (value == this.user.password) || this.$t('validations.samePassword'),
        sameRepeteadPassword: value => (value == this.user.repeteadPassword) || this.$t('validations.samePassword'),
      }
    }
  }
})