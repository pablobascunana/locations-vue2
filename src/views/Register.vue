<template>
  
  <v-form ref="registerForm" @keyup.enter="doRegister" v-model="isFormRegisterValid">
    <v-card :elevation="10" rounded="0">
      <v-card-title class="justify-center">{{ $t('register.title') }}</v-card-title>
      <v-card-text class="pt-0 pb-2">
        <v-row>
          <v-col xl="4" lg="4" md="6" sm="6" cols="12">
            <v-text-field
              id="userName"
              :label="$t('register.userName')"
              type="text"
              v-model="user.userName"
              :rules="[formRules.required, formRules.maxLength]"
              autocomplete="off"
            ></v-text-field>
          </v-col>
          <v-col xl="4" lg="4" md="6" sm="6" cols="12">
            <v-text-field
              id="name"
              :label="$t('register.name')"
              type="text"
              v-model="user.name"
              :rules="[formRules.required, formRules.maxLength]"
              autocomplete="off"
            ></v-text-field>
          </v-col>
          <v-col xl="4" lg="4" md="6" sm="6" cols="12">
            <v-text-field
              id="lastName"
              :label="$t('register.lastName')"
              type="text"
              v-model="user.lastName"
              :rules="[formRules.required, formRules.maxLength]"
              autocomplete="off"
            ></v-text-field>
          </v-col>
          <v-col xl="4" lg="4" md="6" sm="6" cols="12">
            <v-text-field
              id="email"
              :label="$t('register.email')"
              type="text"
              v-model="user.email"
              :rules="[formRules.required, formRules.validEmail]"
              autocomplete="off"
            ></v-text-field>
          </v-col>
          <v-col xl="4" lg="4" md="6" sm="6" cols="12">
            <v-text-field
              id="password"
              :label="$t('register.password')"
              :type="showPassword ? 'text' : 'password'"
              :append-icon="showPassword ? 'mdi-eye' : 'mdi-eye-off'"
              v-model="user.password"
              :rules="[formRules.required, formRules.sameRepeteadPassword, formRules.maxLength]"
              autocomplete="off"
              @click:append="showPassword = !showPassword"
            ></v-text-field>
          </v-col>
          <v-col xl="4" lg="4" md="6" sm="6" cols="12">
            <v-text-field
              id="repeteadPassword"
              :label="$t('register.repeteadPassword')"
              :type="showRepeatedPassword ? 'text' : 'password'"
              :append-icon="showRepeatedPassword ? 'mdi-eye' : 'mdi-eye-off'"
              v-model="user.repeteadPassword"
              :rules="[formRules.required, formRules.samePassword, formRules.maxLength]"
              autocomplete="off"
              @click:append="showRepeatedPassword = !showRepeatedPassword"
            ></v-text-field>
          </v-col>
        </v-row>
      </v-card-text>
      <v-card-actions class="justify-end">
        <v-btn
          id="back" 
          :elevation="0"
          :ripple="false"
          @click.prevent="goToLogin"
        >{{ $t('generic.buttons.back') }}
        </v-btn>
        <v-btn
          id="register"
          :disabled="!isFormRegisterValid"
          :elevation="0"
          @click.prevent="doRegister"
        >{{ $t('generic.buttons.register') }}
        </v-btn>
      </v-card-actions>
    </v-card>
  </v-form>
</template>

<script>
import Api from '@/api/services';
export default {
  name: "Register", 
  data() {
    return {
      isFormRegisterValid: false,
      user: {
        userName: '',
        name: '',
        lastName: '',
        email: '',
        password: '',
        repeteadPassword: '',
      },
      showPassword: false,
      showRepeatedPassword: false
    }
  },
  watch: { 
    user: {
      handler(value) {
        if (value.password === value.repeteadPassword) {
          this.$refs.registerForm.resetValidation();
        }
      },
      deep: true
    }
  },
  methods: {
    async doRegister() {
      try {
        const user = {...this.user};
        delete user.repeteadPassword;
        await Api.register(user);
        this.EventBus.$emit('showSnackBar', { 
          snackBarMessage: `${this.$t('register.messages.success1')} ${this.user.userName} ${this.$t('register.messages.success2')}`
        });
        this.goToLogin();
      } catch(error) {
        this.EventBus.$emit('showSnackBar', { 
          snackBarMessage: this.checkError(error.response.data),
          snackBarColor: 'error'
        });
      }
    },
    goToLogin() {
      this.$router.push({ name: 'login'});
    },
    checkError(errorMessage) {
      if (errorMessage.includes('userName')) {
        return `${this.$t('register.messages.userNameError')} ${this.user.userName} ${this.$t('register.messages.exists')}`;
      } else if (errorMessage.includes('email')) {
        return `${this.$t('register.messages.emailError')} ${this.user.email} ${this.$t('register.messages.exists')}`;
      }
      return `${this.$t('register.messages.error')} ${this.user.userName}`;      
    }
  }
}
</script>
