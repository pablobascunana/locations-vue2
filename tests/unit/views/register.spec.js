import { createLocalVue, shallowMount } from '@vue/test-utils'
import VueRouter from 'vue-router'
import Vuetify from 'vuetify'

import '@/mixins/formRules'
import Register from '@/views/Register.vue'

import * as registerMock from '../mocks/register.mock.js'

const localVue = createLocalVue();
localVue.use(VueRouter);
const router = new VueRouter();
let vuetify = new Vuetify();

describe('Register', () => {
    let wrapper;
    beforeEach(() => {
        wrapper = shallowMount(Register, {
            localVue,
            vuetify,
            router,
            mocks: {
                $t: () => {'some specific text in any language'}
            }
        });
    });

    it("Should match snapshot", () => {
        expect(wrapper).toMatchSnapshot()    
    });

    it('Renders a vue instance', () => {
        expect(wrapper.isVueInstance()).toBe(true);
    });

    it('Click in back button', async () => {
        const button = wrapper.find('#back');
        await button.trigger('click');
        expect(button.attributes().disabled).toBe(undefined);
    });

    it('Register button has disabled property', async () => { 
        const button = wrapper.find('#register');
        await button.trigger('click');
        expect(button.attributes().disabled).toBe('true');
    });

    it('Call watch', () => {
        wrapper.vm.user = registerMock.USER_DATA;
        wrapper.vm.$refs.registerForm.resetValidation = () => true;
    });

    it('Call to check error function', () => {
        const checkErrorFn = jest.spyOn(wrapper.vm, 'checkError');
        expect(wrapper.vm.checkError('userName')).toContain(wrapper.vm.user.userName);
        expect(wrapper.vm.checkError('email')).toContain(wrapper.vm.user.email);
        expect(wrapper.vm.checkError('something')).toContain(wrapper.vm.user.userName);
        expect(checkErrorFn).toBeCalled();
    });
});