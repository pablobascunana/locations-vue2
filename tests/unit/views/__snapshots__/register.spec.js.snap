// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Register Should match snapshot 1`] = `
<v-form>
  <v-card elevation="10" rounded="0">
    <v-card-title class="justify-center"></v-card-title>
    <v-card-text class="pt-0 pb-2">
      <v-row>
        <v-col xl="4" lg="4" md="6" sm="6" cols="12">
          <v-text-field id="userName" type="text" rules="(value) => {
          /* istanbul ignore next */
          cov_1bs8gdva0b.f[1]++;
          cov_1bs8gdva0b.s[2]++;
          return (
            /* istanbul ignore next */
            (cov_1bs8gdva0b.b[0][0]++, !!value) ||
            /* istanbul ignore next */
            (cov_1bs8gdva0b.b[0][1]++, this.$t('validations.required'))
          );
        },(value) => {
          /* istanbul ignore next */
          cov_1bs8gdva0b.f[2]++;
          cov_1bs8gdva0b.s[3]++;
          return (
            /* istanbul ignore next */
            (cov_1bs8gdva0b.b[1][0]++, value.length < 101) ||
            /* istanbul ignore next */
            (cov_1bs8gdva0b.b[1][1]++, this.$t('validations.maxLength'))
          );
        }" autocomplete="off"></v-text-field>
        </v-col>
        <v-col xl="4" lg="4" md="6" sm="6" cols="12">
          <v-text-field id="name" type="text" rules="(value) => {
          /* istanbul ignore next */
          cov_1bs8gdva0b.f[1]++;
          cov_1bs8gdva0b.s[2]++;
          return (
            /* istanbul ignore next */
            (cov_1bs8gdva0b.b[0][0]++, !!value) ||
            /* istanbul ignore next */
            (cov_1bs8gdva0b.b[0][1]++, this.$t('validations.required'))
          );
        },(value) => {
          /* istanbul ignore next */
          cov_1bs8gdva0b.f[2]++;
          cov_1bs8gdva0b.s[3]++;
          return (
            /* istanbul ignore next */
            (cov_1bs8gdva0b.b[1][0]++, value.length < 101) ||
            /* istanbul ignore next */
            (cov_1bs8gdva0b.b[1][1]++, this.$t('validations.maxLength'))
          );
        }" autocomplete="off"></v-text-field>
        </v-col>
        <v-col xl="4" lg="4" md="6" sm="6" cols="12">
          <v-text-field id="lastName" type="text" rules="(value) => {
          /* istanbul ignore next */
          cov_1bs8gdva0b.f[1]++;
          cov_1bs8gdva0b.s[2]++;
          return (
            /* istanbul ignore next */
            (cov_1bs8gdva0b.b[0][0]++, !!value) ||
            /* istanbul ignore next */
            (cov_1bs8gdva0b.b[0][1]++, this.$t('validations.required'))
          );
        },(value) => {
          /* istanbul ignore next */
          cov_1bs8gdva0b.f[2]++;
          cov_1bs8gdva0b.s[3]++;
          return (
            /* istanbul ignore next */
            (cov_1bs8gdva0b.b[1][0]++, value.length < 101) ||
            /* istanbul ignore next */
            (cov_1bs8gdva0b.b[1][1]++, this.$t('validations.maxLength'))
          );
        }" autocomplete="off"></v-text-field>
        </v-col>
        <v-col xl="4" lg="4" md="6" sm="6" cols="12">
          <v-text-field id="email" type="text" rules="(value) => {
          /* istanbul ignore next */
          cov_1bs8gdva0b.f[1]++;
          cov_1bs8gdva0b.s[2]++;
          return (
            /* istanbul ignore next */
            (cov_1bs8gdva0b.b[0][0]++, !!value) ||
            /* istanbul ignore next */
            (cov_1bs8gdva0b.b[0][1]++, this.$t('validations.required'))
          );
        },(value) => {
          /* istanbul ignore next */
          cov_1bs8gdva0b.f[3]++;
          cov_1bs8gdva0b.s[4]++;
          return (
            /* istanbul ignore next */
            (cov_1bs8gdva0b.b[2][0]++, /(([^<>()[\\]\\\\.,;:\\s@&quot;]+(\\.[^<>()[\\]\\\\.,;:\\s@&quot;]+)*)|(&quot;.+&quot;))@(([[0-9]{1,3}\\.[0-9]{1,3}\\.[0-9]{1,3}\\.[0-9]{1,3}\\])|(([a-zA-Z\\-0-9]+\\.)+[a-zA-Z]{2,}))$/.test(value)) ||
            /* istanbul ignore next */
            (cov_1bs8gdva0b.b[2][1]++, this.$t('validations.validEmail'))
          );
        }" autocomplete="off"></v-text-field>
        </v-col>
        <v-col xl="4" lg="4" md="6" sm="6" cols="12">
          <v-text-field id="password" type="password" append-icon="mdi-eye-off" rules="(value) => {
          /* istanbul ignore next */
          cov_1bs8gdva0b.f[1]++;
          cov_1bs8gdva0b.s[2]++;
          return (
            /* istanbul ignore next */
            (cov_1bs8gdva0b.b[0][0]++, !!value) ||
            /* istanbul ignore next */
            (cov_1bs8gdva0b.b[0][1]++, this.$t('validations.required'))
          );
        },(value) => {
          /* istanbul ignore next */
          cov_1bs8gdva0b.f[5]++;
          cov_1bs8gdva0b.s[6]++;
          return (
            /* istanbul ignore next */
            (cov_1bs8gdva0b.b[4][0]++, value == this.user.repeteadPassword) ||
            /* istanbul ignore next */
            (cov_1bs8gdva0b.b[4][1]++, this.$t('validations.samePassword'))
          );
        },(value) => {
          /* istanbul ignore next */
          cov_1bs8gdva0b.f[2]++;
          cov_1bs8gdva0b.s[3]++;
          return (
            /* istanbul ignore next */
            (cov_1bs8gdva0b.b[1][0]++, value.length < 101) ||
            /* istanbul ignore next */
            (cov_1bs8gdva0b.b[1][1]++, this.$t('validations.maxLength'))
          );
        }" autocomplete="off"></v-text-field>
        </v-col>
        <v-col xl="4" lg="4" md="6" sm="6" cols="12">
          <v-text-field id="repeteadPassword" type="password" append-icon="mdi-eye-off" rules="(value) => {
          /* istanbul ignore next */
          cov_1bs8gdva0b.f[1]++;
          cov_1bs8gdva0b.s[2]++;
          return (
            /* istanbul ignore next */
            (cov_1bs8gdva0b.b[0][0]++, !!value) ||
            /* istanbul ignore next */
            (cov_1bs8gdva0b.b[0][1]++, this.$t('validations.required'))
          );
        },(value) => {
          /* istanbul ignore next */
          cov_1bs8gdva0b.f[4]++;
          cov_1bs8gdva0b.s[5]++;
          return (
            /* istanbul ignore next */
            (cov_1bs8gdva0b.b[3][0]++, value == this.user.password) ||
            /* istanbul ignore next */
            (cov_1bs8gdva0b.b[3][1]++, this.$t('validations.samePassword'))
          );
        },(value) => {
          /* istanbul ignore next */
          cov_1bs8gdva0b.f[2]++;
          cov_1bs8gdva0b.s[3]++;
          return (
            /* istanbul ignore next */
            (cov_1bs8gdva0b.b[1][0]++, value.length < 101) ||
            /* istanbul ignore next */
            (cov_1bs8gdva0b.b[1][1]++, this.$t('validations.maxLength'))
          );
        }" autocomplete="off"></v-text-field>
        </v-col>
      </v-row>
    </v-card-text>
    <v-card-actions class="justify-end">
      <v-btn id="back" elevation="0" text="">
      </v-btn>
      <v-btn id="register" color="primary" disabled="true" elevation="0">
      </v-btn>
    </v-card-actions>
  </v-card>
</v-form>
`;

exports[`Register should match snapshot 1`] = `
<v-form>
  <v-card elevation="10" rounded="0">
    <v-card-title class="justify-center"></v-card-title>
    <v-card-text class="pt-0 pb-2">
      <v-row>
        <v-col xl="4" lg="4" md="6" sm="6" cols="12">
          <v-text-field id="userName" type="text" rules="(value) => {
          /* istanbul ignore next */
          cov_1bs8gdva0b.f[1]++;
          cov_1bs8gdva0b.s[2]++;
          return (
            /* istanbul ignore next */
            (cov_1bs8gdva0b.b[0][0]++, !!value) ||
            /* istanbul ignore next */
            (cov_1bs8gdva0b.b[0][1]++, this.$t('validations.required'))
          );
        },(value) => {
          /* istanbul ignore next */
          cov_1bs8gdva0b.f[2]++;
          cov_1bs8gdva0b.s[3]++;
          return (
            /* istanbul ignore next */
            (cov_1bs8gdva0b.b[1][0]++, value.length < 101) ||
            /* istanbul ignore next */
            (cov_1bs8gdva0b.b[1][1]++, this.$t('validations.maxLength'))
          );
        }" autocomplete="off"></v-text-field>
        </v-col>
        <v-col xl="4" lg="4" md="6" sm="6" cols="12">
          <v-text-field id="name" type="text" rules="(value) => {
          /* istanbul ignore next */
          cov_1bs8gdva0b.f[1]++;
          cov_1bs8gdva0b.s[2]++;
          return (
            /* istanbul ignore next */
            (cov_1bs8gdva0b.b[0][0]++, !!value) ||
            /* istanbul ignore next */
            (cov_1bs8gdva0b.b[0][1]++, this.$t('validations.required'))
          );
        },(value) => {
          /* istanbul ignore next */
          cov_1bs8gdva0b.f[2]++;
          cov_1bs8gdva0b.s[3]++;
          return (
            /* istanbul ignore next */
            (cov_1bs8gdva0b.b[1][0]++, value.length < 101) ||
            /* istanbul ignore next */
            (cov_1bs8gdva0b.b[1][1]++, this.$t('validations.maxLength'))
          );
        }" autocomplete="off"></v-text-field>
        </v-col>
        <v-col xl="4" lg="4" md="6" sm="6" cols="12">
          <v-text-field id="lastName" type="text" rules="(value) => {
          /* istanbul ignore next */
          cov_1bs8gdva0b.f[1]++;
          cov_1bs8gdva0b.s[2]++;
          return (
            /* istanbul ignore next */
            (cov_1bs8gdva0b.b[0][0]++, !!value) ||
            /* istanbul ignore next */
            (cov_1bs8gdva0b.b[0][1]++, this.$t('validations.required'))
          );
        },(value) => {
          /* istanbul ignore next */
          cov_1bs8gdva0b.f[2]++;
          cov_1bs8gdva0b.s[3]++;
          return (
            /* istanbul ignore next */
            (cov_1bs8gdva0b.b[1][0]++, value.length < 101) ||
            /* istanbul ignore next */
            (cov_1bs8gdva0b.b[1][1]++, this.$t('validations.maxLength'))
          );
        }" autocomplete="off"></v-text-field>
        </v-col>
        <v-col xl="4" lg="4" md="6" sm="6" cols="12">
          <v-text-field id="email" type="text" rules="(value) => {
          /* istanbul ignore next */
          cov_1bs8gdva0b.f[1]++;
          cov_1bs8gdva0b.s[2]++;
          return (
            /* istanbul ignore next */
            (cov_1bs8gdva0b.b[0][0]++, !!value) ||
            /* istanbul ignore next */
            (cov_1bs8gdva0b.b[0][1]++, this.$t('validations.required'))
          );
        },(value) => {
          /* istanbul ignore next */
          cov_1bs8gdva0b.f[3]++;
          cov_1bs8gdva0b.s[4]++;
          return (
            /* istanbul ignore next */
            (cov_1bs8gdva0b.b[2][0]++, /(([^<>()[\\]\\\\.,;:\\s@&quot;]+(\\.[^<>()[\\]\\\\.,;:\\s@&quot;]+)*)|(&quot;.+&quot;))@(([[0-9]{1,3}\\.[0-9]{1,3}\\.[0-9]{1,3}\\.[0-9]{1,3}\\])|(([a-zA-Z\\-0-9]+\\.)+[a-zA-Z]{2,}))$/.test(value)) ||
            /* istanbul ignore next */
            (cov_1bs8gdva0b.b[2][1]++, this.$t('validations.validEmail'))
          );
        }" autocomplete="off"></v-text-field>
        </v-col>
        <v-col xl="4" lg="4" md="6" sm="6" cols="12">
          <v-text-field id="password" type="password" append-icon="mdi-eye-off" rules="(value) => {
          /* istanbul ignore next */
          cov_1bs8gdva0b.f[1]++;
          cov_1bs8gdva0b.s[2]++;
          return (
            /* istanbul ignore next */
            (cov_1bs8gdva0b.b[0][0]++, !!value) ||
            /* istanbul ignore next */
            (cov_1bs8gdva0b.b[0][1]++, this.$t('validations.required'))
          );
        },(value) => {
          /* istanbul ignore next */
          cov_1bs8gdva0b.f[5]++;
          cov_1bs8gdva0b.s[6]++;
          return (
            /* istanbul ignore next */
            (cov_1bs8gdva0b.b[4][0]++, value == this.user.repeteadPassword) ||
            /* istanbul ignore next */
            (cov_1bs8gdva0b.b[4][1]++, this.$t('validations.samePassword'))
          );
        },(value) => {
          /* istanbul ignore next */
          cov_1bs8gdva0b.f[2]++;
          cov_1bs8gdva0b.s[3]++;
          return (
            /* istanbul ignore next */
            (cov_1bs8gdva0b.b[1][0]++, value.length < 101) ||
            /* istanbul ignore next */
            (cov_1bs8gdva0b.b[1][1]++, this.$t('validations.maxLength'))
          );
        }" autocomplete="off"></v-text-field>
        </v-col>
        <v-col xl="4" lg="4" md="6" sm="6" cols="12">
          <v-text-field id="repeteadPassword" type="password" append-icon="mdi-eye-off" rules="(value) => {
          /* istanbul ignore next */
          cov_1bs8gdva0b.f[1]++;
          cov_1bs8gdva0b.s[2]++;
          return (
            /* istanbul ignore next */
            (cov_1bs8gdva0b.b[0][0]++, !!value) ||
            /* istanbul ignore next */
            (cov_1bs8gdva0b.b[0][1]++, this.$t('validations.required'))
          );
        },(value) => {
          /* istanbul ignore next */
          cov_1bs8gdva0b.f[4]++;
          cov_1bs8gdva0b.s[5]++;
          return (
            /* istanbul ignore next */
            (cov_1bs8gdva0b.b[3][0]++, value == this.user.password) ||
            /* istanbul ignore next */
            (cov_1bs8gdva0b.b[3][1]++, this.$t('validations.samePassword'))
          );
        },(value) => {
          /* istanbul ignore next */
          cov_1bs8gdva0b.f[2]++;
          cov_1bs8gdva0b.s[3]++;
          return (
            /* istanbul ignore next */
            (cov_1bs8gdva0b.b[1][0]++, value.length < 101) ||
            /* istanbul ignore next */
            (cov_1bs8gdva0b.b[1][1]++, this.$t('validations.maxLength'))
          );
        }" autocomplete="off"></v-text-field>
        </v-col>
      </v-row>
    </v-card-text>
    <v-card-actions class="justify-end">
      <v-btn id="back" elevation="0" text="">
      </v-btn>
      <v-btn id="register" color="primary" disabled="true" elevation="0">
      </v-btn>
    </v-card-actions>
  </v-card>
</v-form>
`;
